### PR TITLE
Refactor ParquetExec in preparation for implementing parallel scans for statistics

### LINKED ISSUE #896: Refactor ParquetExec::try_from_files in preparation for making it parallel
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
The logic in `ParquetExec::try_from_files` is not written in a way that makes it easy to parallize.

**Describe the solution you'd like**
Refactor logic to make it easier to apply threading in a separate PR.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes #896.

 # Rationale for this change

Refactor in preparation for making partition scans run in parallel.

# What changes are included in this PR?

Refactor to move some logic from ParquetExec down to ParquetPartition

# Are there any user-facing changes?

No.

### CODE REVIEW

File: datafusion/src/physical_plan/parquet.rs
Comment (andygrove) [inline datafusion/src/physical_plan/parquet.rs:180]: The plan is to use `tokio::spawn` here but this requires making more methods async so I wanted to tackle that as a separate PR.

File: datafusion/src/physical_plan/parquet.rs
Comment (alamb) [inline datafusion/src/physical_plan/parquet.rs:393]: I wonder if having a `Vec<FilenameAndSchema>` would be clearer than two parallel arrays (and then they could not get out of sync either)

File: datafusion/src/physical_plan/parquet.rs
Comment (alamb) [inline datafusion/src/physical_plan/parquet.rs:625]: do you also need to truncate `schemas` as well?

**Review by alamb — APPROVED** [review]
The idea looks very good to me @andygrove  👍 

This may conflict with https://github.com/apache/arrow-datafusion/pull/811/files from @yjshen

I think the comment about `truncate`ing the schemas should probably be addressed before merging, but I don't see any bug that it would cause at the present time, so I would be fine with this PR going in as is.

File: datafusion/src/physical_plan/parquet.rs
Comment (houqp) [reply datafusion/src/physical_plan/parquet.rs:393]: It seems like the `PartitionedFile` abstraction proposed by @yjshen  in https://github.com/apache/arrow-datafusion/pull/811/files#diff-72f3a52c56e83e00d8c605d461f092617a3c205619376bb373069c662f9cfc93R54 would help solve this problem?

File: datafusion/src/physical_plan/parquet.rs
Comment (houqp) [reply datafusion/src/physical_plan/parquet.rs:625]: looks like it.

**Comment by yjshen** [comment]
Ah, seems I was addressing the same issue (parallel stats listing) pointed out by @rdettai here 😂: https://github.com/apache/arrow-datafusion/pull/811/files#diff-72f3a52c56e83e00d8c605d461f092617a3c205619376bb373069c662f9cfc93R189-R223

**Comment by andygrove** [comment]
@yjshen I have changed this PR to a draft and will hold off on working on this for now and will review your PR when I have time - probably at the weekend. It looks like you are farther along than I was with this.

**Comment by andygrove** [comment]
Closing this in favor of the work happening in https://github.com/apache/arrow-datafusion/pull/811

**Pull request closed without merging**
